File: Data Structures & Algorithms/eating-bananas/submission-3.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int l = 1;
        int r = 0;

        for(int p : piles){
            r = Math.max(r, p);
        }

        int res = r;

        while(l <= r){
            int m = l + (r - l) / 2;
            if(canEat(piles, h, m)){
                res = m;
                r = m - 1;
               
            }else{
                l = m + 1;
            }
        }

        return res;
        
    }

    private boolean canEat(int[] piles, int h, int m){
        int currH = 0;
        for(int p : piles){
            currH += Math.ceil((double) p / m);

            if(currH > h){
                return false;
            }
        }

        return true;
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-3.js
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    maxAreaOfIsland(grid) {
        const row = grid.length;
        const col = grid[0].length;

        let max = 0;

        const dfs = (r, c) => {
            if (r < 0 || c < 0 || r == row || c == col || grid[r][c] == "#" || grid[r][c] == 0) {
                return 0;
            }

            grid[r][c] = "#";
            const res = 1 + dfs(r + 1, c) + dfs(r - 1, c) + dfs(r, c + 1) + dfs(r, c - 1);

            return res;
        };

        for(let r = 0; r < row; r++){
            for(let c = 0; c < col; c++){
                if(grid[r][c] == 1){
                    max = Math.max(max, dfs(r, c));
                }
            }
        }

        return max;
    }
}
